### PR TITLE
[WiP] Version 3.2.0 - Feature Release

### PR DESCRIPTION
**Pull request by SamBrishes**
Hello,

as discussed via mail, I'm preparing the first pull request for the upcoming release (and future development) of the Mall plugin. ~~There are still 2 issues that I definitely want to push into the new minor version:~~

- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/102
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/423
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/684
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/963
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/1001
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/1029
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/1032
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/1076
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/1081
- [x] https://github.com/OFFLINE-GmbH/oc-mall-plugin/issues/1085
- [x] Implement Payment Fee / Discount, write test cases
- [x] Check / re-implement getFilteredTaxes (traits/FilteredTaxes)
- [x] Testing backward-compatibility with OCv2

## Notes

@tobias-kuendig 

1. All tests went green on OC v2.2 and OC v3.5 (with exception of skipped "Transferred carts get merged" test).
2. I had to decrease the required PHP version to keep support with OC v2.2.
3. The documentation pages has been moved to ./resources/docs/
4. I commented -out all `updates/version.yaml` lines.

## Highlights

<details>
<summary><h4>New Navigation Icons</h4></summary>

Using custom vectors graphics instead of FontAwesome / October Icons. Created with Affinity Designer, the source files are located in the`resources` directory.

image

image
</details>
<details>
<summary><h4>Enabled State on different models</h4></summary>

The following models use the new `IsState` model trait, which introduces an additional `is_enabled` column and controls the `is_default` one in a common manner as well. This means that the individual records can now be de- and activated separately and do not have to be deleted.

- Currency
- PriceCategory
- Tax
- Payment Methods
- Shipping Methods
- Order States

Example:
image
</details>
<details>
<summary><h4>Introducing the new PriceBag TotalsCalculator</h4></summary>

In the first step, the new `PriceBag` calcultion collector works - more or less - just alongslide the 
current TotalsCalculator implementation (even if PriceBag already performs most of the calculations 
within TotalsCalculator). However, during the upcoming releases, `PriceBag` will be further expanded, 
designed to be more bullet-proof and performant until it will fully replace the whole TotalsCalculcator 
bundle (including all related classes). In version 4.0.0 at the latest, but probably sooner.

</details>
<details>
<summary><h4>Introducing a new Documentation Page</h4></summary>

The new documentation site is now based on VitePress, the (in)direct successor to VuePress, and 
therefore also comes with a completely new look. While most of the documentation pages are still a 
work in progress, progress is being made at breakneck speed.

</details>
<details>
<summary><h4>Updated console commands / New Seeders</h4></summary>

The artisan console-commands have been revised and renamed, and adapted to the new seeders. The DB Seeders itself has been split up into the initial core-records (`OFFLINE\Mall\Updates\Seeders\CoreSeeder`) and the demonstration ones (`OFFLINE\Mall\Updates\Seeders\DemoSeeder`) and can also be called using Octobers' native `plugin:seed` command. The demonstration data has also been translated into german.

```
 mall
  mall:check            Check if your setup is complete
  mall:index            Recreate the general product index
  mall:purge            Purge all customer and order related data
  mall:seed             Seed the Mall related database records
```

```
> php artisan mall:seed --help

Description:
  Seed the Mall related database records.

Usage:
  mall:seed [options]

Options:
      --force            Don't ask before erasing all records
  -d, --with-demo        Insert demonstration records, such as products
  -l, --locale[=LOCALE]  Force a specific locale for the seeded records
```

</details>

- \+ Fixed Test-Suite / Testing environment
- \+ General Clean-Up & Fixes

## Changes

Related Issues:

- https://github.com/RatMD/oc-mall-plugin/issues/7
- https://github.com/RatMD/oc-mall-plugin/issues/21
- https://github.com/RatMD/oc-mall-plugin/issues/24
- https://github.com/RatMD/oc-mall-plugin/issues/25
- https://github.com/RatMD/oc-mall-plugin/issues/26

### CODE REVIEW

**Comment by SamBrishes** [comment]
Small Update:

I'm currently writing a new `PriceBag`-called replacement for the TotalsCalculator classes, which should fix the current discount issue but should also provide additional support for upcoming features.

However, 22 tests (from the TotalsCalculatorTest class) are still ahead.

**Comment by SamBrishes** [comment]
Some progress done...

```sh
> composer test -- --filter 'OFFLINE\\Mall\\Tests\\Classes'
PHPUnit 9.6.16 by Sebastian Bergmann and contributors.

Price Bag (OFFLINE\Mall\Tests\Classes\Pricing\PriceBag)
 ✔ Calculate price for single product
 ✔ Calculate price for single product with service option
 ✔ Calculate price using multiple taxes on service option
 ✔ Calculate price on multiple products
 ✔ Calculate taxes correctly
 ✔ Calculate taxes correctly after discount applied
 ✔ Calculate correctly after discount results in zero amount
 ✔ Calculate multiple products with different taxes and discount
 ✔ Calculate excluded taxes correctly
 ✔ Calculate excluded taxes correctly after discount applied
 ✔ Calculate shipping costs
 ✔ Calculate forced shipping costs
 ✔ Calculate taxes correctly when excluded with shipping
 ✔ Calculate taxes correctly when excluded with quantity
 ✔ Calculate detailed taxes on excluded with shipping
 ✔ Calculate product weight total
 ✔ Calculate product weight total using variants
 ✔ Calculate shipping cost using rate
 ✔ Calculate variant cost
 ✔ Calculate custom fields costs
 ✔ Calculate custom fields fallback costs
 ✔ Calculate fixed discount
 ✔ Calculate rate discount
 ✔ Calculate rate discount always on base price
 ✔ Calculate alternate shipping price discount
 ✔ Calculate fixed discount only when total is reached
 ✔ Calculate rate discount only when total is reached
 ✔ Calculate alternate shipping price discount only when total is reached
 ✔ Calculate fixed discount only when product is in cart
 ✔ Calculate rate discount only when product is in cart
 ✔ Calculate alternate shipping price discount only when product is in cart

Totals Calculator (OFFLINE\Mall\Tests\Classes\Totals\TotalsCalculator)
 ✔ It works for a single product
 ✔ It works for a single product with service options
 ✔ It works for multiple products with service options
 ✔ It works for multiple products
 ✔ It calculates taxes included
 ✔ It calculates taxes included on amount after discount applied
 ✔ It calculates taxes included on zero amount after discount applied
 ✔ It calculates taxes with different taxes and discount
 ✔ It calculates taxes excluded
 ✔ It calculates taxes excluded with discount
 ✔ It calculates taxes excluded on amount after discount applied
 ✔ It calculates shipping cost
 ✔ It calculates enforced shipping cost
 ✔ It calculates taxes
 ✔ It calculates taxes with quantity
 ✔ It consolidates taxes
 ✔ It calculates detailed taxes
 ✔ It calculates weight total
 ✔ It calculates weight total with variants
 ✔ It calculates shipping cost with special rates
 ✔ It calculates variant cost
 ✔ It calculates custom fields cost
 ✔ It calculates custom field fallback cost
 ✔ It applies fixed discounts
 ✔ It applies rate discounts
 ✔ It applies rate discounts always to base price
 ✔ It applies alternate shipping price discounts
 ✔ It applies fixed amount discount only when given total is reached
 ✔ It applies rate discounts only when given total is reached
 ✔ It applies alternate shipping price discounts only when given total is reached
 ✔ It applies fixed amount discount only when needed product is in cart
 ✔ It applies rate discounts only when needed product is in cart
 ✔ It applies alternate shipping price discounts only when needed product is in cart

Time: 05:23.353, Memory: 68.00 MB

OK (64 tests, 184 assertions)
```

some small tasks still ahead...

- [ ] Implement Payment Fee / Discount, write test cases
- [ ] Check / re-implement getFilteredTaxes (traits/FilteredTaxes)
- [ ] Custom MoneyValue instead of Money to provide consistent access to values
- [ ] Get rid of current applyDiscount() method

Current progress is available in the `next` branch!

**Comment by SamBrishes** [comment]
@tobias-kuendig Have you found some time to take a look at this new release? 

I would like to take the next steps, but wanted to wait for this one first.

Thanks.

File: updates/version.yaml [hunk 1]
```diff
@@ -1,490 +1,525 @@
...
+#    - 'Fixed out of stock message even if stock was available, thanks to @xyz1123581321.'
+#    - 'Fixed wrong indexed value key on PropertyValue using color types, thanks to @toome123.'
+#    - 'Fixed wrong taxes calculation when discounts has been applied.'
+#    - 'Fixed wrong taxes calculation when different discounts has been applied to products and shipping costs.'
+#    - 'Fixed bad SQL error when addresses has been added to a customer profile on the backend.'
+#    - 'Fixed duplication mess on products with variants.'
+#    - 'Fixed possibility to take-over email address on CustomerProfile component, thanks to @cyril-design.'
+#    - 'Fixed products / variants disappear on first page view after editing a product.'
+#    - 'Removed legacy v1 code.'
+#    - 030_01-update_system_plugin_history.php
```
Comment (tobias-kuendig) [inline updates/version.yaml:517]: What is the purpose of this migration? It currently fails on my test installation:

> In 030_01-update_system_plugin_history.php line 74:
>
>   Migration file "030_01-add_product_variants_sort_order.php" on path "/user/code/oc-mall-plugin/updates" does not exist.

File: classes/utils/DefaultMoney.php [hunk 2]
```diff
@@ -23,6 +23,7 @@ public function format(?float $value, $product = null, ?Currency $currency = nul
     {
         $currency = $currency ?? Currency::activeCurrency();
 
+        $orig = $value;
```
Comment (tobias-kuendig) [inline classes/utils/DefaultMoney.php:26]: This looks unused

```suggestion
```

**Review by tobias-kuendig — CHANGES_REQUESTED** [review]
Just a few minor things!

- [x] grafik

The Reviews Icon seems to be missing.

**Comment by SamBrishes** [comment]
> Just a few minor things!
> 
> grafik
> 
> 
> The Reviews Icon seems to be missing.

Nice, this seems to be an October v3.6 "bug". The icon name has been changed from `icon-star-half-full` to `icon-star-half`. I have to check if the new name also works under v2, but it should be fixed quickly.

https://docs.octobercms.com/3.x/element/available-icons.html#available-october-icons

v3.6 (left) vs. v3.5 (right):

image

**Comment by SamBrishes** [comment]
Since OC v3.6 has been released literally 2 ago, I'll test the plugin again against the new version. Changing to draft again. (I'm currently busy, may will have some left time on the weekend.)

**Comment by tobias-kuendig** [comment]
@SamBrishes I've seen you have added your `localize` client to the repo. This is a great idea and simplifies the translation work. 

Is there any way we can include it as a proper dependency instead of putting the whole project in this repo? I'd like to not bloat the repo even more ‒ especially considering that 99% if plugin installs won't do translation work.

**Comment by SamBrishes** [comment]
Sorry for the late reply, too much overtime too less free time at the moment.

> Is there any way we can include it as a proper dependency instead of putting the whole project in this repo? I'd like to not bloat the repo even more ‒ especially considering that 99% if plugin installs won't do translation work.

@tobias-kuendig Of course. I have published the translation tool in [a separate repository](https://github.com/RatMD/oc-translator) and integrated it via `.npmrc` as a `package.json` dependency.

File: updates/version.yaml (hunk 1 above)
Comment (SamBrishes) [reply updates/version.yaml:547]: Finally fixed this issue. Looks like v2 has been developed alongside v3, which resulted in different migration histories and (gladly just) one additional update-file. (Which, by the way, led to many errors when migrating from Mall v2.0.12 or higher to master / v3.)

However, to answer your question. This change was supposed to improve the overall view of the available migration files by using groups and a file index as a prefix. This pattern allows to simple identify which update-files belongs together, which simplifies backtracking migration / migrate:refresh issues.

Of course, I can undo this change if you wish.

**Comment by SamBrishes** [comment]
Hello @tobias-kuendig ,

- Tested with OctoberCMS v2.2.34 + Migration (v2 -> v3)
- Tested with OctoberCMS v3.6.17

Second try.

**Comment by tobias-kuendig** [comment]
I'll give this a review ASAP, thanks @SamBrishes!

> I guess it makes sense using the next branch as WiP and push future releases to develop instead, since develop is linked to the OC composer repository?

Absolutely!

**Review by tobias-kuendig — APPROVED** [review]

**Comment by tobias-kuendig** [comment]
@SamBrishes the MR looks good to me! Great work, congratulations!

Let me know if you have any further changes to make. If not, I'll make the release ready this week.

**Comment by SamBrishes** [comment]
Thanks, 

no further changes will follow in this pull request. I think smaller steps are more appropriate after this huge one.

**Pull request merged**